### PR TITLE
Improving ScalaDoc for ExecutionContexts

### PR DESCRIPTION
**Pull request by viktorklang**
Improving the guidelines for ExecutionContext and blocking.

### CODE REVIEW

File: src/library/scala/concurrent/ExecutionContext.scala [hunk 1]
```diff
@@ -25,20 +25,18 @@ import scala.annotation.implicitNotFound
...
- * logically-related section of code) will make a decision about
- * which `ExecutionContext` to use. That is, you might want to avoid
- * hardcoding `scala.concurrent.ExecutionContext.Implicits.global` all
- * over the place in your code.
- * One approach is to add `(implicit ec: ExecutionContext)`
- * to methods which need an `ExecutionContext`. Then import a specific
- * context in one place for the entire application or module,
+ * implicit `ExecutionContext`, application developers should carefully
+ * consider where they want to set execution policy;
+ * ideally, one place per application—or per logically-related section of code—
```
Comment (janekdb) [inline src/library/scala/concurrent/ExecutionContext.scala]: Over use of hyphens?

File: src/library/scala/concurrent/ExecutionContext.scala [hunk 2]
```diff
@@ -25,20 +25,18 @@ import scala.annotation.implicitNotFound
...
+ * will make a decision about which `ExecutionContext` to use.
+ * That is, you will mostly want to avoid hardcoding, especially vid an import,
```
Comment (janekdb) [inline src/library/scala/concurrent/ExecutionContext.scala]: via

File: src/library/scala/concurrent/ExecutionContext.scala (hunk 1 above)
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: 'tis not a hyphen, brother. It's an m-dash.

File: src/library/scala/concurrent/ExecutionContext.scala (hunk 2 above)
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: 👍

File: src/library/scala/concurrent/ExecutionContext.scala [hunk 3]
```diff
@@ -25,20 +25,18 @@ import scala.annotation.implicitNotFound
...
+ * `scala.concurrent.ExecutionContext.Implicits.global`.
+ * The recommended approach is to add `(implicit ec: ExecutionContext)` to methods,
+ * or class constructor parameters, which need an `ExecutionContext`.
+ * Then locally import, or define a local implicit val with, a specific
```
Comment (janekdb) [inline src/library/scala/concurrent/ExecutionContext.scala]: Less compact but easier to read?

> Then locally import a specific `ExecutionContext` in one place for the entire application or module passing it implicitly to individual methods.  Alternatively define a local implicit val with the required `ExecutionContext`.

File: src/library/scala/concurrent/ExecutionContext.scala [hunk 4]
```diff
@@ -111,9 +109,18 @@ object ExecutionContext {
...
-   * [[https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html#availableProcessors-- available processors]].
+   * The default `ExecutionContext` implementation is backed by a work-stealing thread pool.
+   * It can be configured via the following [[scala.sys.SystemProperties]]:
+   *
+   * `scala.concurrent.context.minThreads` = defaults to "1"
+   * `scala.concurrent.context.numThreads` = defaults to "x1" (i.e. the current number of available processors * 1)
+   * `scala.concurrent.context.maxThreads` = defaults to "x1" (i.e. the current number of available processors * 1)
+   * `scala.concurrent.context.maxExtraThreads` = defaults to "256"
+   *
+   * The pool size of threads is then `numThreads` bounded by `minThreads` on the lower end and `maxThread` on the high end.
```
Comment (janekdb) [inline src/library/scala/concurrent/ExecutionContext.scala]: maxThreads

Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: 👍

File: src/library/scala/concurrent/package.scala [hunk 5]
```diff
@@ -140,17 +159,19 @@ package concurrent {
...
+   * which will cause performance degradation, and possibly, deadlock issues.
    */
   object Await {
     /**
      * Await the "completed" state of an `Awaitable`.
      *
      * Although this method is blocking, the internal use of [[scala.concurrent.blocking blocking]] ensures that
-     * the underlying [[ExecutionContext]] is prepared to properly manage the blocking.
+     * the underlying [[ExecutionContext]] is given an opportunity to properly manage the blocking.
+     * WARNING: It is strongly discouraged to supply lenghty timeouts, as the progress of the calling thread will be
```
Comment (janekdb) [inline src/library/scala/concurrent/package.scala]: lengthy

File: src/library/scala/concurrent/ExecutionContext.scala (hunk 3 above)
Comment (viktorklang) [reply src/library/scala/concurrent/ExecutionContext.scala]: Nice. Reworked the punctuation

File: src/library/scala/concurrent/package.scala [hunk 6]
```diff
@@ -140,17 +159,19 @@ package concurrent {
...
+     * suspended (blocked) until either the `Awaitable` has a ready or the timeout expires.
```
Comment (janekdb) [inline src/library/scala/concurrent/package.scala]: "is ready" or "has a result" to match below?

File: src/library/scala/concurrent/package.scala [hunk 7]
```diff
@@ -172,7 +193,10 @@ package concurrent {
      * Await and return the result (of type `T`) of an `Awaitable`.
      *
      * Although this method is blocking, the internal use of [[scala.concurrent.blocking blocking]] ensures that
-     * the underlying [[ExecutionContext]] to properly detect blocking and ensure that there are no deadlocks.
+     * the underlying [[ExecutionContext]] is given an opportunity to properly manage the blocking.
+     *
+     * WARNING: It is strongly discouraged to supply lenghty timeouts, as the progress of the calling thread will be
```
Comment (janekdb) [inline src/library/scala/concurrent/package.scala]: lengthy

File: src/library/scala/concurrent/package.scala (hunk 6 above)
Comment (viktorklang) [reply src/library/scala/concurrent/package.scala]: replaced with `becomes ready`

File: src/library/scala/concurrent/package.scala (hunk 7 above)
Comment (viktorklang) [reply src/library/scala/concurrent/package.scala]: fixed

**Comment by viktorklang** [comment]
Massive thanks to @janekdb for the proof-reading!

File: src/library/scala/concurrent/package.scala [hunk 8]
```diff
@@ -140,17 +159,20 @@ package concurrent {
...
   object Await {
     /**
      * Await the "completed" state of an `Awaitable`.
      *
      * Although this method is blocking, the internal use of [[scala.concurrent.blocking blocking]] ensures that
-     * the underlying [[ExecutionContext]] is prepared to properly manage the blocking.
+     * the underlying [[ExecutionContext]] is given an opportunity to properly manage the blocking.
+     *
+     * WARNING: It is strongly discouraged to supply lengthy timeouts since the progress of the calling thread will be
+     * suspended—blocked—until either the `Awaitable` becomes ready or the timeout expires.
```
Comment (janekdb) [inline src/library/scala/concurrent/package.scala:175]: `suspended—blocked—until`

Does this mean "..will be suspended and therefore blocked until either .."

Comment (viktorklang) [reply src/library/scala/concurrent/package.scala:175]: Yes and no. I it equivalent to "suspended (blocked) until".

**Comment by SethTisue** [comment]
thanks Viktor and Janek!

**Pull request merged**